runtime/lua/vim/lsp/buf.lua

```diff
@@ -143,6 +143,79 @@ local function select_client(method, on_choice)
   end
 end
 
+--- Utility function to request formatting from attached language servers. Defaults to
+--- requesting formatting from all servers attached to the current buffer.
+---
+--- @param opts table Optional table which holds the following optional fields:
+---     - formatting_options (table):
+---         Can be used to specify FormattingOptions. Some unspecified options will be
+---         automatically derived from the current Neovim options.
+---         @see https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting
+---     - timeout_ms (integer, default 1000):
+---         Time in milliseconds to block for formatting requests. Formatting requests are current
+---         synchronous to prevent editing of the buffer.
+---     - bufnr (number):
+---         Restrict formatting to the clients attached to the given buffer, defaults to the current
+---         buffer (0).
+---     - filter (function):
+---         Restrict formatting to client ids returned by this function. Takes as argument the
+---         clients attached to the buffer (default to the current buffer) and returns a list
+---         of client objects from which to request formatting.
+---
+---         Example:
+---         <pre>
+---         -- Never request typescript-language-server for formatting
+---         vim.lsp.buf.format {
+---           filter = function(clients)
+---             return vim.tbl_filter(
+---               function(client) return client.name ~= "tsserver" end,
+---               clients
+---             )
+---           end
+---         }
+---         </pre>
+---
+---     - id (number):
+---         Restrict formatting to the client with ID (client.id) matching this field.
+---     - name (string):
+---         Restrict formatting to the client with name (client.name) matching this field.
+
+function M.format(opts)
+  opts = opts or {}
+  local bufnr = opts.bufnr or vim.api.nvim_get_current_buf()
+  local clients = vim.lsp.buf_get_clients(bufnr)
+
+  if opts.filter then
+    clients = opts.filter(clients)
+  elseif opts.id then
+    clients = vim.tbl_filter(
+      function(client) return client.id == opts.id end,
+      clients
+    )
+  elseif opts.name then
+    clients = vim.tbl_filter(
+      function(client) return client.name == opts.name end,
+      clients
+    )
+  end
+
+  clients = vim.tbl_filter(
+    function(client) return client.supports_method("textDocument/formatting") end,
+    clients
+  )
+
+  local timeout_ms =  opts.timeout_ms or 1000
+  for _, client in pairs(clients) do
```

Comment:
I think there should be some notification if no client matched. Otherwise users have a "nothing happens" experience which isn't ideal.